File: api-discovery.mjs
#!/usr/bin/env node
/**
 * Backend API Discovery Script
 * Discovers and catalogs backend endpoints for admin catalog/components
 */

import http from 'http';
import https from 'https';
import { URL } from 'url';

const BACKEND = 'http://192.168.1.35:3000';
const EMAIL = 'admin@tudominio.com';
const PASSWORD = 'Admin@12345Local';
const TIMEOUT = 5000;

// Helper function to make HTTP requests
function makeRequest(method, path, body = null, headers = {}) {
  return new Promise((resolve, reject) => {
    const targetUrl = new URL(path, BACKEND);
    const options = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers
      },
      timeout: TIMEOUT
    };

    const protocol = targetUrl.protocol === 'https:' ? https : http;
    
    const req = protocol.request(targetUrl, options, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const parsed = data ? JSON.parse(data) : null;
          resolve({
            status: res.statusCode,
            headers: res.headers,
            body: parsed,
            raw: data
          });
        } catch (e) {
          resolve({
            status: res.statusCode,
            headers: res.headers,
            body: null,
            raw: data,
            parseError: e.message
          });
        }
      });
    });

    req.on('error', reject);
    req.on('timeout', () => {
      req.destroy();
      reject(new Error('Request timeout'));
    });

    if (body) {
      req.write(JSON.stringify(body));
    }
    req.end();
  });
}

function logSection(title) {
  console.log(`\n${'='.repeat(60)}`);
  console.log(`${title}`);
  console.log('='.repeat(60));
}

function logSuccess(msg) {
  console.log(`✓ ${msg}`);
}

function logError(msg) {
  console.log(`✗ ${msg}`);
}

function logInfo(msg) {
  console.log(`ℹ ${msg}`);
}

async function discoverBackend() {
  logSection('BACKEND API DISCOVERY');
  logInfo(`Target: ${BACKEND}`);
  logInfo(`Credentials: ${EMAIL}`);

  // Step 1: Test connectivity
  console.log('\n[1] Testing backend connectivity...');
  try {
    const res = await makeRequest('GET', '/');
    if (res.status) {
      logSuccess(`Backend is UP (Status: ${res.status})`);
    }
  } catch (err) {
    logError(`Backend not reachable: ${err.message}`);
    return;
  }

  // Step 2: Authenticate
  let token = null;
  console.log('\n[2] Attempting authentication...');
  
  const authEndpoints = ['/auth/login', '/api/auth/login', '/api/authentication/login'];
  
  for (const endpoint of authEndpoints) {
    try {
      logInfo(`Trying ${endpoint}...`);
      const res = await makeRequest('POST', endpoint, { email: EMAIL, password: PASSWORD });
      
      if (res.status === 200 || res.status === 201) {
        logSuccess(`Authentication successful at ${endpoint} (Status: ${res.status})`);
        logInfo(`Response keys: ${Object.keys(res.body || {}).join(', ')}`);
        
        // Extract token from various possible fields
        token = res.body?.token || res.body?.access_token || res.body?.accessToken || res.body?.data?.token;
        
        if (token) {
          logSuccess(`Token received: ${token.substring(0, 40)}...`);
          break;
        } else {
          logInfo(`No token field found in response`);
          logInfo(`Full response: ${JSON.stringify(res.body, null, 2)}`);
        }
      } else {
        logInfo(`${endpoint} returned status ${res.status}`);
      }
    } catch (err) {
      logInfo(`${endpoint} failed: ${err.message}`);
    }
  }

  if (!token) {
    logError('Failed to authenticate - cannot proceed');
    return;
  }

  // Step 3: Probe catalog/component endpoints
  console.log('\n[3] Probing catalog/component endpoints...');
  
  const endpoints = [
    '/components',
    '/catalog',
    '/products',
    '/admin/components',
    '/admin/catalog',
    '/admin/products',
    '/api/components',
    '/api/catalog',
    '/api/products',
    '/api/admin/components',
    '/api/admin/catalog',
    '/api/admin/products',
    '/v1/components',
    '/v1/admin/components'
  ];

  const authHeaders = {
    'Authorization': `Bearer ${token}`
  };

  const discoveredEndpoints = [];

  for (const endpoint of endpoints) {
    logInfo(`\nTesting ${endpoint}...`);
    
    try {
      const res = await makeRequest('GET', endpoint, null, authHeaders);
      
      if (res.status === 200 || res.status === 201) {
        logSuccess(`${endpoint} - Status ${res.status}`);
        
        const endpointInfo = {
          method: 'GET',
          path: endpoint,
          status: res.status,
          responseKeys: res.body ? Object.keys(res.body) : [],
          isList: Array.isArray(res.body),
          itemSampleKeys: null,
          itemSample: null
        };

        if (Array.isArray(res.body) && res.body.length > 0) {
          endpointInfo.itemSampleKeys = Object.keys(res.body[0]);
          endpointInfo.itemSample = res.body[0];
          logInfo(`  Response is array with ${res.body.length} items`);
          logInfo(`  First item keys: ${Object.keys(res.body[0]).join(', ')}`);
        } else if (res.body && typeof res.body === 'object') {
          logInfo(`  Response keys: ${Object.keys(res.body).join(', ')}`);
          
          // Check if data is nested in a data property
          if (res.body.data && Array.isArray(res.body.data) && res.body.data.length > 0) {
            endpointInfo.itemSampleKeys = Object.keys(res.body.data[0]);
            endpointInfo.itemSample = res.body.data[0];
            logInfo(`  Nested data array with ${res.body.data.length} items`);
            logInfo(`  First item keys: ${Object.keys(res.body.data[0]).join(', ')}`);
          }
        }

        discoveredEndpoints.push(endpointInfo);
      } else {
        logInfo(`  Status: ${res.status} (skipped)`);
      }
    } catch (err) {
      logInfo(`  Error: ${err.message}`);
    }
  }

  // Step 4: Try OPTIONS to discover allowed methods
  console.log('\n[4] Testing OPTIONS method on discovered endpoints...');
  
  for (const endpoint of discoveredEndpoints.slice(0, 3)) {
    try {
      const res = await makeRequest('OPTIONS', endpoint.path, null, authHeaders);
      if (res.headers['allow']) {
        logInfo(`${endpoint.path} allows: ${res.headers['allow']}`);
      }
    } catch (err) {
      logInfo(`OPTIONS ${endpoint.path}: ${err.message}`);
    }
  }

  // Step 5: Test query parameters on first discovered endpoint
  if (discoveredEndpoints.length > 0) {
    console.log('\n[5] Testing pagination/filters on first endpoint...');
    const endpoint = discoveredEndpoints[0];
    
    const queryTests = [
      { path: `${endpoint.path}?page=1&limit=10`, desc: 'Pagination' },
      { path: `${endpoint.path}?limit=5`, desc: 'Limit only' },
      { path: `${endpoint.path}?sort=id&order=asc`, desc: 'Sorting' },
      { path: `${endpoint.path}?search=test`, desc: 'Search' },
      { path: `${endpoint.path}?filter=active`, desc: 'Filter' }
    ];

    for (const test of queryTests) {
      try {
        const res = await makeRequest('GET', test.path, null, authHeaders);
        if (res.status === 200) {
          logSuccess(`${test.desc}: ${test.path}`);
        }
      } catch (err) {
        // Silently skip
      }
    }
  }

  // Summary
  logSection('DISCOVERY RESULTS');
  console.log(`\nEndpoints Found: ${discoveredEndpoints.length}`);
  
  if (discoveredEndpoints.length > 0) {
    console.log('\nEndpoints Details:');
    discoveredEndpoints.forEach((ep, i) => {
      console.log(`\n  ${i + 1}. ${ep.path}`);
      console.log(`     Method: ${ep.method}`);
      console.log(`     Status: ${ep.status}`);
      console.log(`     Response Keys: ${ep.responseKeys.join(', ')}`);
      if (ep.isList) {
        console.log(`     Type: List (Array)`);
        if (ep.itemSampleKeys) {
          console.log(`     Item Keys: ${ep.itemSampleKeys.join(', ')}`);
          console.log(`     Sample Item: ${JSON.stringify(ep.itemSample, null, 6).split('\n').slice(0, 8).join('\n')}`);
        }
      }
    });
  }

  console.log('\n' + '='.repeat(60));
}

// Run discovery
discoverBackend().catch(err => {
  logError(`Fatal error: ${err.message}`);
  process.exit(1);
});

File: api_discovery.py
#!/usr/bin/env python3
"""
Backend API Discovery Script
Discovers and catalogs backend endpoints for admin catalog/components
"""

import json
import urllib.request
import urllib.error
import sys
from typing import Optional, Dict, Any

BACKEND = "http://192.168.1.35:3000"
EMAIL = "admin@tudominio.com"
PASSWORD = "Admin@12345Local"
TIMEOUT = 5

def make_request(method: str, path: str, body: Optional[Dict] = None, headers: Optional[Dict] = None) -> Dict[str, Any]:
    """Make an HTTP request and return the response."""
    if headers is None:
        headers = {}
    
    url = f"{BACKEND}{path}"
    headers['Content-Type'] = 'application/json'
    headers['User-Agent'] = 'APIDiscovery/1.0'
    
    if body:
        data = json.dumps(body).encode('utf-8')
    else:
        data = None
    
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    
    try:
        with urllib.request.urlopen(req, timeout=TIMEOUT) as response:
            response_data = response.read().decode('utf-8')
            status = response.status
            try:
                parsed = json.loads(response_data) if response_data else None
            except json.JSONDecodeError:
                parsed = None
            
            return {
                'status': status,
                'body': parsed,
                'raw': response_data
            }
    except urllib.error.HTTPError as e:
        response_data = e.read().decode('utf-8')
        try:
            parsed = json.loads(response_data) if response_data else None
        except json.JSONDecodeError:
            parsed = None
        
        return {
            'status': e.code,
            'body': parsed,
            'raw': response_data
        }
    except urllib.error.URLError as e:
        return {
            'status': 0,
            'error': str(e),
            'body': None,
            'raw': ''
        }
    except Exception as e:
        return {
            'status': 0,
            'error': str(e),
            'body': None,
            'raw': ''
        }

def main():
    print("\n" + "="*70)
    print("BACKEND API ENDPOINT DISCOVERY")
    print("="*70)
    print(f"Backend: {BACKEND}")
    print(f"Credentials: {EMAIL}\n")
    
    # Step 1: Test connectivity
    print("[1] Testing backend connectivity...")
    res = make_request('GET', '/')
    if res['status'] > 0:
        print(f"✓ Backend is reachable (Status: {res['status']})\n")
    else:
        print(f"✗ Backend connection failed: {res.get('error', 'Unknown error')}\n")
        return
    
    # Step 2: Authenticate
    token = None
    print("[2] Attempting authentication...")
    
    res = make_request('POST', '/auth/login', {
        'email': EMAIL,
        'password': PASSWORD
    })
    
    if res['status'] in (200, 201):
        print(f"✓ Auth successful at /auth/login (Status: {res['status']})")
        if res['body']:
            print(f"  Response keys: {', '.join(res['body'].keys())}")
            token = (res['body'].get('token') or 
                    res['body'].get('access_token') or 
                    res['body'].get('accessToken') or
                    (res['body'].get('data', {}).get('token') if isinstance(res['body'].get('data'), dict) else None))
            
            if token:
                print(f"✓ Token received: {token[:40]}...")
            else:
                print("✗ No token found in response")
                print(f"  Full response: {json.dumps(res['body'], indent=2)}")
    else:
        print(f"✗ Auth failed at /auth/login (Status: {res['status']})")
        print(f"  Response: {res['raw']}")
    
    if not token:
        print("\n✗ Cannot proceed without authentication token\n")
        return
    
    # Step 3: Probe endpoints
    print("\n[3] Probing catalog/component endpoints...\n")
    
    endpoints = [
        '/components',
        '/catalog',
        '/products',
        '/admin/components',
        '/admin/catalog',
        '/admin/products',
        '/api/components',
        '/api/catalog',
        '/api/products',
        '/api/admin/components',
        '/api/admin/catalog',
        '/api/admin/products',
        '/v1/components',
        '/v1/admin/components',
        '/users'
    ]
    
    auth_headers = {
        'Authorization': f'Bearer {token}'
    }
    
    discovered = []
    
    for endpoint in endpoints:
        res = make_request('GET', endpoint, headers=auth_headers)
        
        if res['status'] in (200, 201):
            print(f"✓ {endpoint}")
            print(f"  Method: GET | Status: {res['status']}")
            
            ep_info = {
                'method': 'GET',
                'path': endpoint,
                'status': res['status'],
                'body': res['body'],
                'response_keys': list(res['body'].keys()) if isinstance(res['body'], dict) else [],
                'is_list': isinstance(res['body'], list)
            }
            
            if isinstance(res['body'], list):
                if len(res['body']) > 0:
                    ep_info['item_count'] = len(res['body'])
                    ep_info['item_keys'] = list(res['body'][0].keys()) if isinstance(res['body'][0], dict) else []
                    print(f"  Type: Array ({len(res['body'])} items)")
                    if res['body'][0]:
                        print(f"  Item keys: {', '.join(ep_info['item_keys'])}")
                else:
                    print(f"  Type: Empty Array")
            elif isinstance(res['body'], dict):
                print(f"  Type: Object")
                print(f"  Keys: {', '.join(res['body'].keys())}")
                
                if 'data' in res['body'] and isinstance(res['body']['data'], list) and len(res['body']['data']) > 0:
                    ep_info['nested_data'] = True
                    ep_info['data_keys'] = list(res['body']['data'][0].keys()) if isinstance(res['body']['data'][0], dict) else []
                    print(f"  Nested data array: {len(res['body']['data'])} items")
                    if res['body']['data'][0]:
                        print(f"  Data item keys: {', '.join(ep_info['data_keys'])}")
            
            discovered.append(ep_info)
        elif res['status'] == 401:
            print(f"⊘ {endpoint} (Status: 401 Unauthorized)")
        elif res['status'] == 403:
            print(f"⊘ {endpoint} (Status: 403 Forbidden)")
        elif res['status'] == 404:
            print(f"⊘ {endpoint} (Status: 404 Not Found)")
        else:
            print(f"⊘ {endpoint} (Status: {res['status']})")
        
        print("")
    
    # Summary
    print("="*70)
    print(f"SUMMARY: Found {len(discovered)} working endpoints\n")
    
    for i, ep in enumerate(discovered, 1):
        print(f"\n{i}. {ep['path']}")
        print(f"   Method: {ep['method']}")
        print(f"   Status: {ep['status']}")
        print(f"   Response Type: {'Array' if ep['is_list'] else 'Object'}")
        
        if ep['is_list'] and ep.get('item_keys'):
            print(f"   Item Structure: {', '.join(ep['item_keys'])}")
            if ep['body'] and len(ep['body']) > 0 and isinstance(ep['body'][0], dict):
                print(f"   Sample Item:")
                sample_str = json.dumps(ep['body'][0], indent=4)
                for line in sample_str.split('\n')[:15]:
                    print(f"     {line}")
        elif ep.get('nested_data'):
            print(f"   Data Array Items: {ep.get('item_count', '?')}")
            if ep.get('data_keys'):
                print(f"   Item Structure: {', '.join(ep['data_keys'])}")
    
    print("\n" + "="*70 + "\n")

if __name__ == "__main__":
    main()
